manager/integration/tests/test_zone.py

```diff
@@ -301,3 +330,611 @@ def test_replica_zone_anti_affinity(client, core_api, volume_name, k8s_node_zone
+            break
+
+        time.sleep(RETRY_INTERVAL)
+    assert z1_r_count == 6
+    assert z2_r_count == 0
+    assert z3_r_count == 0
+
+    client.update(n2, allowScheduling=True)
+
+    for _ in range(RETRY_COUNTS):
+        z1_r_count = get_zone_replica_count(
+            client, volume_name, ZONE1, chk_running=True)
+        z2_r_count = get_zone_replica_count(
+            client, volume_name, ZONE2, chk_running=True)
+        z3_r_count = get_zone_replica_count(
+            client, volume_name, ZONE3, chk_running=True)
+
+        if z1_r_count == z2_r_count == 3 and z3_r_count == 0:
+            break
+
+        time.sleep(RETRY_INTERVAL)
+    assert z1_r_count == 3
+    assert z2_r_count == 3
+    assert z3_r_count == 0
+
+    client.update(n3, allowScheduling=True)
+
+    for _ in range(RETRY_COUNTS):
+        z1_r_count = get_zone_replica_count(
+            client, volume_name, ZONE1, chk_running=True)
+        z2_r_count = get_zone_replica_count(
+            client, volume_name, ZONE2, chk_running=True)
+        z3_r_count = get_zone_replica_count(
+            client, volume_name, ZONE3, chk_running=True)
+
+        if z1_r_count == z2_r_count == z3_r_count == 2:
+            break
+
+        time.sleep(RETRY_INTERVAL)
+    assert z1_r_count == 2
+    assert z2_r_count == 2
+    assert z3_r_count == 2
+
+    reset_longhorn_node_zone(client, core_api)
+
+
+def test_replica_auto_balance_zone_best_effort_with_data_locality(
+        client, core_api, volume_name, pod):  # NOQA
+    """
+    Background:
+    Given set `replica-soft-anti-affinity` to `true`.
+    And set `replica-zone-soft-anti-affinity` to `true`.
+    And set `default-data-locality` to `best-effort`.
+    And set `replicaAutoBalance` to `best-effort`.
+    And set node-1 to zone-1.
+        set node-2 to zone-1.
+        set node-3 to zone-2.
+    And create volume with 2 replicas.
+    And create pv for volume.
+    And create pvc for volume.
+
+    Scenario: replica auto-balance zones with best-effort should not
+              remove pod local replicas when data locality is enabled
+              (best-effort).
+
+    Given create and wait pod on <node>.
+    And disable scheduling and evict node-3.
+    And count replicas on each nodes.
+    And 1 replica running on <pod-node>.
+        1 replica running on <duplicate-node>.
+        0 replica running on node-3.
+
+    When enable scheduling for node-3.
+    Then count replicas on each nodes.
+    And 1 replica running on <pod-node>.
+        0 replica running on <duplicate-node>.
+        1 replica running on node-3.
+    And count replicas in each zones.
+    And 1 replica running in zone-1.
+        1 replica running in zone-2.
+
+    When sleep 10 seconds, to ensure no addition scheduling is happening.
+    Then count replicas on each nodes.
+    And 1 replica running on <pod-node>.
+        0 replica running on <duplicate-node>.
+        1 replica running on node-3.
+
+    And delete pod.
```

Comment:
You can launch the pod on another node of zone1 then see if the replica in zone1 will be migrated to the new node.

Reply:
Already included the in the scenario with the loop through of deploying to `node-1` > `node-2` > `node-1`.